src/lib.rs

```diff
@@ -1,506 +1,632 @@
+    /// internally.
+    fn log(&self, record: &LogRecord);
 }
 
-#[doc(hidden)]
+/// The location of a log message.
+#[derive(Copy, Clone, Debug)]
 pub struct LogLocation {
+    /// The module path of the message.
     pub module_path: &'static str,
+    /// The source file containing the message.
     pub file: &'static str,
+    /// The line containing the message.
     pub line: usize,
 }
 
-impl Copy for LogLocation {}
-
-/// Tests whether a given module's name is enabled for a particular level of
-/// logging. This is the second layer of defense about determining whether a
-/// module's log statement should be emitted or not.
-#[doc(hidden)]
-pub fn mod_enabled(level: u32, module: &str) -> bool {
-    static INIT: Once = ONCE_INIT;
-    INIT.call_once(init);
-
-    // It's possible for many threads are in this function, only one of them
-    // will perform the global initialization, but all of them will need to check
-    // again to whether they should really be here or not. Hence, despite this
-    // check being expanded manually in the logging macro, this function checks
-    // the log level again.
-    if level > unsafe { LOG_LEVEL } { return false }
-
-    // This assertion should never get tripped unless we're in an at_exit
-    // handler after logging has been torn down and a logging attempt was made.
-    assert!(unsafe { !DIRECTIVES.is_null() });
-
-    enabled(level, module, unsafe { (*DIRECTIVES).iter() })
-}
-
-fn enabled(level: u32,
-           module: &str,
-           iter: slice::Iter<directive::LogDirective>)
-           -> bool {
-    // Search for the longest match, the vector is assumed to be pre-sorted.
-    for directive in iter.rev() {
-        match directive.name {
-            Some(ref name) if !module.starts_with(name.as_slice()) => {},
-            Some(..) | None => {
-                return level <= directive.level
-            }
-        }
+/// A token providing read and write access to the global maximum log level
+/// filter.
+///
+/// The maximum log level is used as an optimization to avoid evaluating log
+/// messages that will be ignored by the logger. Any message with a level
+/// higher than the maximum log level filter will be ignored. A logger should
+/// make sure to keep the maximum log level filter in sync with its current
+/// configuration.
+#[allow(missing_copy_implementations)]
+pub struct MaxLogLevelFilter(());
+
+impl fmt::Debug for MaxLogLevelFilter {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "MaxLogLevelFilter")
     }
-    level <= DEFAULT_LOG_LEVEL
 }
 
-/// Initialize logging for the current process.
+impl MaxLogLevelFilter {
+    /// Gets the current maximum log level filter.
+    pub fn get(&self) -> LogLevelFilter {
+        max_log_level()
+    }
+
+    /// Sets the maximum log level.
+    pub fn set(&self, level: LogLevelFilter) {
+        MAX_LOG_LEVEL_FILTER.store(level as usize, Ordering::Relaxed)
+    }
+}
```

Comment:
This seems like a bit of an odd API to only be used by the initialization function. Is the purpose here to provide a guarantee of one-time initialization, but allow that initialization to be optional?